Source/ROLib/Modules/ROLAnimateEngineHeat.cs

```diff
@@ -0,0 +1,219 @@
+                locateSolverEngines();
+        }
+
+        public void Update()
+        {
+            if (HighLogic.LoadedSceneIsFlight)
+                updateHeat();
+        }
+
+        private void locateEngineModule()
+        {
+            engineModule = null;
+            engineModule = part.GetComponents<ModuleEngines>().FirstOrDefault(x => x.engineID == engineID);
+            if (engineModule == null)
+                MonoBehaviour.print("ERROR: Could not locate engine by ID: " + engineID + " for part: " + part +
+                                    " for ROLAnimateEngineHeat.  This will cause errors during gameplay.  Setting engine to first engine module (if present)");
+
+            engineModule ??= part.GetComponent<ModuleEngines>();
+        }
+
+        private void locateAnimatedTransforms()
+        {
+            List<Renderer> renderers = new List<Renderer>();
+            Transform[] animatedTransforms = part.transform.ROLFindChildren(meshName);
+            foreach (var animatedTransform in animatedTransforms)
+            {
+                renderers.AddRange(animatedTransform.GetComponentsInChildren<Renderer>(false));
+            }
+            animatedRenderers = renderers.ToArray();
+            if (animatedRenderers == null || animatedRenderers.Length == 0) { print("ERROR: Could not locate any emissive meshes for name: " + meshName); }
+        }
+
+        private void locateSolverEngines()
+        {
+            if (ROLModInterop.getSolverEngineModule(part, engineID) is ModuleEngines solverEnginesModule)
+            {
+                MethodInfo getter = ROLModInterop.getSolverEngineTempProperty().GetGetMethod();
+
+                getSolverChamberTemp = (Func<double>) Delegate.CreateDelegate(typeof(Func<double>), solverEnginesModule, getter);
+            }
+        }
+
+        private void updateHeat()
+        {
+            if (engineModule == null) { return; }
+
+            float emissivePercent = 0f;
+
+            if (!useThrottle && !useSolverEngines)
+            {
+                //add heat from engine
+                if (engineModule.EngineIgnited && !engineModule.flameout && engineModule.currentThrottle > 0)
+                {
+                    float throttle = engineModule.currentThrottle;
+                    float heatIn = heatAccumulationCurve.Evaluate(throttle) * heatOutput * TimeWarp.fixedDeltaTime;
+                    currentHeat += heatIn;
+                }
+
+                //dissipate heat
+                float heatPercent = currentHeat / maxHeat;
+                if (currentHeat > 0f)
+                {
+                    float heatOut = heatDissipationCurve.Evaluate(heatPercent) * heatDissipation * TimeWarp.fixedDeltaTime;
+                    if (heatOut > currentHeat) { heatOut = currentHeat; }
+                    currentHeat -= heatOut;
+                }
+                if (currentHeat > maxStoredHeat) { currentHeat = maxStoredHeat; }
```

Comment:
Think this is my last comment.  Physics calculations like this do belong in FixedUpdate [or something that runs on that schedule], and running them in Update() will actually be wrong.

You could move just this part of the physics calculation into FixedUpdate().
Better would be to actually start a MonoBehaviour.StartCoroutine() or use the TimingManager <https://kerbalspaceprogram.com/api/class_timing_manager.html>'s FixedUpdateAdd() method to only start your physics coroutine IF you will have physics to do, ie:
In OnStart:
```
if (!useThrottle && !useSolverEngines && engineModule) StartCoroutine(EngineHeatCalc())
```

```
private System.Collections.IEnumerator EngineHeatCalc()
{
     while (engineModule is ModuleEngines)   // Probably unnecessary, while(true) would be okay
    {
        // Do all the physics calcs.
    }
}

Reply:
OK, sounds good. Just out of interest, is there any specific reason this should run in FixedUpdate(), or is it just good practice to put calculations like that in FixedUpdate?

Reply:
Also, is there some kind of list of what can be returned by `yield return`, or is it just `WaitForSeconds()` and `WaitForFixedUpdate()`?

Reply:
Physics stuff should run in FixedUpdate, because you get the [semi?-]fixed step size and one call per physics tick.  Update() runs once per graphics frame, which is a variable number of calls per game-second.  You need the consistent step size to do things like how you use `TimeWarp.fixedDeltaTime`.

There are several others, see https://docs.unity3d.com/ScriptReference/WaitForEndOfFrame.html   though I'm not sure off-hand how you learn about the different ones available.  (Or create others that inherit from YieldInstruction?  Getting beyond my experience.)